Autopilot/SensorFusion/Src/SensorFusion.cpp

```diff
@@ -0,0 +1,445 @@
+
+    float newX[DIM*1];
+    for (int i = 0; i < DIM*1; i++) newX[i] = x[i] + kMultZMinHMultX[i];
+
+    //Defines for updating error covariance
+
+    float I[DIM*DIM] =
+    {
+        1, 0, 0, 0, 0, 0,
+        0, 1, 0, 0, 0, 0,
+        0, 0, 1, 0, 0, 0,
+        0, 0, 0, 1, 0, 0,
+        0, 0, 0, 0, 1, 0,
+        0, 0, 0, 0, 0, 1
+    };
+
+    //Update error covariance: newP = (I - k*h)*p
+
+    float kMultH[DIM*DIM];
+    mul(k, h, kMultH, DIM, NUM_MEASUREMENTS, DIM);
+
+    float IMinKMultH[DIM*DIM];
+    for (int i = 0; i < DIM*DIM; i++) IMinKMultH[i] = I[i] - kMultH[i];
+
+    float newP[DIM*DIM];
+    mul(IMinKMultH, p, newP, DIM, DIM, DIM);
+
+
+    /*Output*/
+
+    Output->altitude = x[0];
+    Output->latitude = x[2];
+    Output->longitude = x[4];
+
+    for (int i = 0; i < DIM*1; i++) iterdata->prevX[i] = newX[i];
+    for (int i = 0; i < DIM*DIM; i++) iterdata->prevP[i] = newP[i];
+
+    return SFError;
+}
+
+SFError_t SF_GenerateNewResult()
+{
+    SFError_t SFError;
+    SFError.errorCode = 0;
+    
+    IMUData_t imuData;
+    GpsData_t GpsData;
+    AltimeterData_t altimeterData;
+    airspeedData_t airspeedData;
+    imuObj->GetResult(imuData);
+    gpsObj->GetResult(&GpsData);
+    //altimeterObj->GetResult(altimeterData);
+    //airspeedObj->GetResult(airspeedData);
+
+    SFAttitudeOutput_t attitudeOutput;
+    SFPositionOutput_t positionOutput;
+
+    SFError = SF_GetAttitude(&attitudeOutput, &imuData, &airspeedData);
+    if(SFError.errorCode != 0) return SFError;
+
+    SFError = SF_GetPosition(&positionOutput, &altimeterData, &GpsData, &imuData, &attitudeOutput, &iterData);
+    if(SFError.errorCode != 0) return SFError;
+
+    SFOutput.pitch = attitudeOutput.pitch;
+    SFOutput.roll = attitudeOutput.roll;
+    SFOutput.yaw = attitudeOutput.yaw;
+
+    return SFError;
+}
+
+SFError_t SF_GetResult(SFOutput_t *output)
+{
+    SFError_t SFError;
+    SFError.errorCode = 0;
+
+    *output = SFOutput;
+
+    return SFError;
+}
+
+IMU_Data_t SF_GetRawIMU()
+{
+    IMUData_t imuData;
+    //imuData = imuObj->getResult();
+
+    IMU_Data_t imuOutput;
+
+    imuOutput.accx = imuData.accx;
+    imuOutput.accy = imuData.accy;
+    imuOutput.accz = imuData.accz;
+    imuOutput.gyrx = imuData.gyrx;
+    imuOutput.gyry = imuData.gyry;
+    imuOutput.gyrz = imuData.gyrz;
+    imuOutput.magx = imuData.magx;
+    imuOutput.magy = imuData.magy;
+    imuOutput.magz = imuData.magz;
+    imuOutput.isDataNew = imuData.isDataNew;
+    imuOutput.sensorStatus = imuData.sensorStatus;
+
+    return imuOutput;
+}
+
+Airspeed_Data_t SF_GetRawAirspeed()
+{
+    airspeedData_t airspeedData;
+    //airspeedData = airspeedObj->getResult();
+
+    Airspeed_Data_t airspeedOutput;
+
+    airspeedOutput.airspeed = airspeedData.airspeed;
+    airspeedOutput.utcTime = airspeedData.utcTime;
+    airspeedOutput.isDataNew = airspeedData.isDataNew;
+    airspeedOutput.sensorStatus = airspeedData.sensorStatus;
+
+    return airspeedOutput;
```

Comment:
ditto (though for this one I understand you are still waiting on the driver). 

What I will say tho is this: there really is no need for you to create the `airspeedOutput` struct. Just return the `airspeedData` struct directly.